ext/drreg/drreg.c

```diff
@@ -441,6 +727,59 @@ drreg_event_bb_insert_late(void *drcontext, void *tag, instrlist_t *bb, instr_t
     }
 
     /* Before each app read, or at end of bb, restore spilled registers to app values: */
+#ifdef X86
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++) {
+        restored_for_simd_read[SIMD_IDX(reg)] = false;
+        if (!pt->simd_reg[SIMD_IDX(reg)].native) {
+            if (drmgr_is_last_instr(drcontext, inst) ||
+                /* This covers reads from all simds, because the applicable range
+                 * resembles zmm, and all other x86 simds are included in zmm.
+                 */
+                instr_reads_from_reg(inst, reg, DR_QUERY_INCLUDE_ALL) ||
+                /* FIXME i#3844: For ymm and zmm support, we're missing support to restore
+                 * upon a partial simd write. For example a write to xmm while zmm is
+                 * clobbered, or a partial write with an evex mask.
+                 */
+                /* i#1954: for complex bbs we must restore before the next app instr */
+                (!pt->simd_reg[SIMD_IDX(reg)].in_use &&
+                 ((pt->bb_has_internal_flow &&
+                   !TEST(DRREG_IGNORE_CONTROL_FLOW, pt->bb_props)) ||
+                  TEST(DRREG_CONTAINS_SPANNING_CONTROL_FLOW, pt->bb_props)))) {
```

Comment:
This complex code is very similar to the GPR code.  Plus, this function is now very large with two copies of each big section of code.  Can they be easily shared?

Reply:
Sharing is not easy because there are nuances that are different enough to make it difficult. We could factor this out into separate functions. @johnfxgalea , I think we can at least do that, is factoring out the simd and the gpr loops in a static helper feasible? I think this would make it more readable.

Reply:
I guess I have the same reaction as before about the similar code and the function size.  If sharing is too hard: please add a comment stating that (or else I'll probably waste some time in the future trying to do it, having forgotten this conversation).

Could we split the function into pieces, though, even if we can't share: that would make it easier to read?

Reply:
There should also be comments about how if a change to one of these blocks with identical logic is made, the other must also be changed to keep them in sync (the complex conditions in those if statements).